icu4c/source/python/icutools/databuilder/filtration.py

```diff
@@ -151,83 +158,86 @@ def match(self, file):
-        # Store the locale as *requested*
-        self.locales_requested.add(locale)
-        # Store the locale and its dependencies as *required*
-        while locale is not None:
-            self.locales_required.add(locale)
-            locale = self._get_parent_locale(locale)
+        # Load the dependency graph from disk
+        self.dependency_data_by_tree = {
+            tree: io.read_locale_deps(tree)
+            for tree in utils.ALL_TREES
+        }
 
     def match(self, file):
+        tree = self._file_to_subdir(file)
+        assert tree is not None
         locale = self._file_to_file_stem(file)
 
         # A locale is *required* if it is *requested* or an ancestor of a
         # *requested* locale.
-        if locale in self.locales_required:
+        if locale in self._locales_required(tree):
             return True
 
         # Resolve include_scripts and include_children.
-        return self._match_recursive(locale)
+        return self._match_recursive(locale, tree)
 
-    def _match_recursive(self, locale):
+    def _match_recursive(self, locale, tree):
         # Base case: return True if we reached a *requested* locale,
         # or False if we ascend out of the locale tree.
         if locale is None:
             return False
-        if locale in self.locales_requested:
+        if locale in self.requested_locales:
             return True
 
         # Check for alternative scripts.
         # This causes sr_Latn to check sr instead of going directly to root.
         if self.include_scripts:
             match = LANGUAGE_SCRIPT_REGEX.match(locale)
-            if match and self._match_recursive(match.group(1)):
+            if match and self._match_recursive(match.group(1), tree):
                 return True
 
         # Check if we are a descendant of a *requested* locale.
         if self.include_children:
-            parent = self._get_parent_locale(locale)
-            if self._match_recursive(parent):
+            parent = self._get_parent_locale(locale, tree)
+            if self._match_recursive(parent, tree):
                 return True
 
         # No matches.
         return False
 
-    @classmethod
-    def _get_parent_locale(cls, locale):
-        if locale in DEPENDENCY_DATA["parents"]:
-            return DEPENDENCY_DATA["parents"][locale]
-        if locale in DEPENDENCY_DATA["aliases"]:
-            return DEPENDENCY_DATA["aliases"][locale]
+    def _get_parent_locale(self, locale, tree):
+        """Gets the parent locale in the given tree, according to dependency data."""
+        dependency_data = self.dependency_data_by_tree[tree]
+        if locale in dependency_data["parents"]:
+            return dependency_data["parents"][locale]
+        if locale in dependency_data["aliases"]:
+            return dependency_data["aliases"][locale]
         if LANGUAGE_ONLY_REGEX.match(locale):
             return "root"
         i = locale.rfind("_")
```

Comment:
Isn't it true at this point that it's a hard failure to not have 'root' here? Everything else should have been dealt with.
Then the comment on this can say "only returns None if the input was "root".